[Bug]: 1.19.0 regression with Kafka

### Module

Kafka

### Testcontainers version

1.19.0

### Using the latest Testcontainers version?

Yes

### Host OS

Linux, Mac OS

### Host Arch

x86, arm, amd64

### Docker version

```shell
Client:
 Cloud integration: v1.0.31
 Version:           23.0.5
 API version:       1.42
 Go version:        go1.19.8
 Git commit:        bc4487a
 Built:             Wed Apr 26 16:12:52 2023
 OS/Arch:           darwin/amd64
 Context:           default

Server: Docker Desktop 4.19.0 (106363)
 Engine:
  Version:          23.0.5
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.8
  Git commit:       94d3ad6
  Built:            Wed Apr 26 16:17:45 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.20
  GitCommit:        2806fc1057397dbaeefbea0e4e17bddfbd388f38
 runc:
  Version:          1.1.5
  GitCommit:        v1.1.5-0-gf19387a
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```


### What happened?

When upgrading from 1.18.3 to 1.19.0 out tests starting a kafka container are failing to start a container.

### Relevant log output

```shell
===> Running preflight checks ... 
    ===> Launching ... 
    ===> Launching kafka ... 
    [2023-08-24 20:26:10,676] INFO Registered kafka:type=kafka.Log4jController MBean (kafka.utils.Log4jControllerRegistration$)
    [2023-08-24 20:26:11,003] INFO Setting -D jdk.tls.rejectClientInitiatedRenegotiation=true to disable client-initiated TLS renegotiation (org.apache.zookeeper.common.X509Util)
    [2023-08-24 20:26:11,047] ERROR Exiting Kafka due to fatal exception (kafka.Kafka$)
    java.lang.IllegalArgumentException: Error creating broker listeners from 'EXTERNAL_PLAINTEXT://localhost:32894,EXTERNAL_SASL_PLAINTEXT://localhost:32892,INTERNAL_PLAINTEXT://kafka-broker.testcontainer.docker:9094,BROKER://kafka-broker.testcontainer.docker:9092': No security protocol defined for listener EXTERNAL_PLAINTEXT
    	at kafka.utils.CoreUtils$.listenerListToEndPoints(CoreUtils.scala:273)
    	at kafka.server.KafkaConfig.effectiveAdvertisedListeners(KafkaConfig.scala:1924)
    	at kafka.server.KafkaConfig.validateValues(KafkaConfig.scala:2033)
    	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:1997)
    	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:1471)
    	at kafka.Kafka$.buildServer(Kafka.scala:67)
    	at kafka.Kafka$.main(Kafka.scala:87)
    	at kafka.Kafka.main(Kafka.scala)
    Caused by: java.lang.IllegalArgumentException: No security protocol defined for listener EXTERNAL_PLAINTEXT
    	at kafka.cluster.EndPoint$.$anonfun$createEndPoint$2(EndPoint.scala:49)
    	at scala.collection.immutable.Map$Map2.getOrElse(Map.scala:307)
    	at kafka.cluster.EndPoint$.securityProtocol$1(EndPoint.scala:49)
    	at kafka.cluster.EndPoint$.createEndPoint(EndPoint.scala:57)
    	at kafka.utils.CoreUtils$.$anonfun$listenerListToEndPoints$6(CoreUtils.scala:270)
    	at scala.collection.StrictOptimizedIterableOps.map(StrictOptimizedIterableOps.scala:100)
    	at scala.collection.StrictOptimizedIterableOps.map$(StrictOptimizedIterableOps.scala:87)
    	at scala.collection.mutable.ArraySeq.map(ArraySeq.scala:37)
    	at kafka.utils.CoreUtils$.listenerListToEndPoints(CoreUtils.scala:270)
```


### Additional Information

_No response_

--- a/modules/kafka/src/test/java/org/testcontainers/containers/KafkaContainerTest.java
+++ b/modules/kafka/src/test/java/org/testcontainers/containers/KafkaContainerTest.java
@@ -1,6 +1,7 @@
 package org.testcontainers.containers;
 
 import com.google.common.collect.ImmutableMap;
+import lombok.SneakyThrows;
 import org.apache.kafka.clients.admin.AdminClient;
 import org.apache.kafka.clients.admin.AdminClientConfig;
 import org.apache.kafka.clients.admin.NewTopic;
@@ -11,8 +12,12 @@ import org.apache.kafka.clients.consumer.KafkaConsumer;
 import org.apache.kafka.clients.producer.KafkaProducer;
 import org.apache.kafka.clients.producer.ProducerConfig;
 import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.config.SaslConfigs;
+import org.apache.kafka.common.errors.SaslAuthenticationException;
+import org.apache.kafka.common.errors.TopicAuthorizationException;
 import org.apache.kafka.common.serialization.StringDeserializer;
 import org.apache.kafka.common.serialization.StringSerializer;
+import org.awaitility.Awaitility;
 import org.junit.Test;
 import org.rnorth.ducttape.unreliables.Unreliables;
 import org.testcontainers.Testcontainers;
@@ -22,10 +27,12 @@ import org.testcontainers.utility.DockerImageName;
 import java.time.Duration;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Properties;
 import java.util.UUID;
 import java.util.concurrent.TimeUnit;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.assertj.core.api.Assertions.tuple;
 
 public class KafkaContainerTest {
@@ -36,6 +43,15 @@ public class KafkaContainerTest {
 
     private static final DockerImageName ZOOKEEPER_TEST_IMAGE = DockerImageName.parse(
         "confluentinc/cp-zookeeper:4.0.0"
+    );
+
+    private final ImmutableMap<String, String> properties = ImmutableMap.of(
+        AdminClientConfig.SECURITY_PROTOCOL_CONFIG,
+        "SASL_PLAINTEXT",
+        SaslConfigs.SASL_MECHANISM,
+        "PLAIN",
+        SaslConfigs.SASL_JAAS_CONFIG,
+        "org.apache.kafka.common.security.plain.PlainLoginModule required username=\"admin\" password=\"admin\";"
     );
 
     @Test
@@ -220,34 +236,166 @@ public class KafkaContainerTest {
         }
     }
 
-    protected void testKafkaFunctionality(String bootstrapServers) throws Exception {
-        testKafkaFunctionality(bootstrapServers, 1, 1);
+    @SneakyThrows
+    @Test
+    public void shouldConfigureAuthenticationWithSaslUsingJaas() {
+        try (
+            KafkaContainer kafka = new KafkaContainer(DockerImageName.parse("confluentinc/cp-kafka:6.2.1"))
+                .withEnv("KAFKA_LISTENER_SECURITY_PROTOCOL_MAP", "PLAINTEXT:SASL_PLAINTEXT,BROKER:SASL_PLAINTEXT")
+                .withEnv("KAFKA_SASL_MECHANISM_INTER_BROKER_PROTOCOL", "PLAIN")
+                .withEnv("KAFKA_LISTENER_NAME_PLAINTEXT_SASL_ENABLED_MECHANISMS", "PLAIN")
+                .withEnv("KAFKA_LISTENER_NAME_BROKER_SASL_ENABLED_MECHANISMS", "PLAIN")
+                .withEnv("KAFKA_LISTENER_NAME_BROKER_PLAIN_SASL_JAAS_CONFIG", getJaasConfig())
+                .withEnv("KAFKA_LISTENER_NAME_PLAINTEXT_PLAIN_SASL_JAAS_CONFIG", getJaasConfig())
+        ) {
+            kafka.start();
+
+            testSecureKafkaFunctionality(kafka.getBootstrapServers());
+        }
     }
 
-    protected void testKafkaFunctionality(String bootstrapServers, int partitions, int rf) throws Exception {
+    @SneakyThrows
+    @Test
+    public void enableSaslWithUnsuccessfulTopicCreation() {
         try (
+            KafkaContainer kafka = new KafkaContainer(DockerImageName.parse("confluentinc/cp-kafka:6.2.1"))
+                .withEnv("KAFKA_LISTENER_SECURITY_PROTOCOL_MAP", "PLAINTEXT:SASL_PLAINTEXT,BROKER:SASL_PLAINTEXT")
+                .withEnv("KAFKA_SASL_MECHANISM_INTER_BROKER_PROTOCOL", "PLAIN")
+                .withEnv("KAFKA_LISTENER_NAME_PLAINTEXT_SASL_ENABLED_MECHANISMS", "PLAIN")
+                .withEnv("KAFKA_LISTENER_NAME_BROKER_SASL_ENABLED_MECHANISMS", "PLAIN")
+                .withEnv("KAFKA_LISTENER_NAME_BROKER_PLAIN_SASL_JAAS_CONFIG", getJaasConfig())
+                .withEnv("KAFKA_LISTENER_NAME_PLAINTEXT_PLAIN_SASL_JAAS_CONFIG", getJaasConfig())
+                .withEnv("KAFKA_AUTHORIZER_CLASS_NAME", "kafka.security.authorizer.AclAuthorizer")
+                .withEnv("KAFKA_SUPER_USERS", "User:admin")
+        ) {
+            kafka.start();
+
             AdminClient adminClient = AdminClient.create(
-                ImmutableMap.of(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers)
-            );
-            KafkaProducer<String, String> producer = new KafkaProducer<>(
                 ImmutableMap.of(
-                    ProducerConfig.BOOTSTRAP_SERVERS_CONFIG,
-                    bootstrapServers,
-                    ProducerConfig.CLIENT_ID_CONFIG,
-                    UUID.randomUUID().toString()
-                ),
+                    AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG,
+                    kafka.getBootstrapServers(),
+                    AdminClientConfig.SECURITY_PROTOCOL_CONFIG,
+                    "SASL_PLAINTEXT",
+                    SaslConfigs.SASL_MECHANISM,
+                    "PLAIN",
+                    SaslConfigs.SASL_JAAS_CONFIG,
+                    "org.apache.kafka.common.security.plain.PlainLoginModule required username=\"test\" password=\"secret\";"
+                )
+            );
+
+            String topicName = "messages-" + UUID.randomUUID();
+            Collection<NewTopic> topics = Collections.singletonList(new NewTopic(topicName, 1, (short) 1));
+
+            Awaitility
+                .await()
+                .untilAsserted(() -> {
+                    assertThatThrownBy(() -> adminClient.createTopics(topics).all().get(30, TimeUnit.SECONDS))
+                        .hasCauseInstanceOf(TopicAuthorizationException.class);
+                });
+        }
+    }
+
+    @SneakyThrows
+    @Test
+    public void enableSaslAndWithAuthenticationError() {
+        String jaasConfig = getJaasConfig();
+        try (
+            KafkaContainer kafka = new KafkaContainer(DockerImageName.parse("confluentinc/cp-kafka:6.2.1"))
+                .withEnv("KAFKA_LISTENER_SECURITY_PROTOCOL_MAP", "PLAINTEXT:SASL_PLAINTEXT,BROKER:SASL_PLAINTEXT")
+                .withEnv("KAFKA_SASL_MECHANISM_INTER_BROKER_PROTOCOL", "PLAIN")
+                .withEnv("KAFKA_LISTENER_NAME_PLAINTEXT_SASL_ENABLED_MECHANISMS", "PLAIN")
+                .withEnv("KAFKA_LISTENER_NAME_BROKER_SASL_ENABLED_MECHANISMS", "PLAIN")
+                .withEnv("KAFKA_LISTENER_NAME_BROKER_PLAIN_SASL_JAAS_CONFIG", jaasConfig)
+                .withEnv("KAFKA_LISTENER_NAME_PLAINTEXT_PLAIN_SASL_JAAS_CONFIG", jaasConfig)
+        ) {
+            kafka.start();
+
+            AdminClient adminClient = AdminClient.create(
+                ImmutableMap.of(
+                    AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG,
+                    kafka.getBootstrapServers(),
+                    AdminClientConfig.SECURITY_PROTOCOL_CONFIG,
+                    "SASL_PLAINTEXT",
+                    SaslConfigs.SASL_MECHANISM,
+                    "PLAIN",
+                    SaslConfigs.SASL_JAAS_CONFIG,
+                    "org.apache.kafka.common.security.plain.PlainLoginModule required username=\"test\" password=\"secretx\";"
+                )
+            );
+
+            String topicName = "messages-" + UUID.randomUUID();
+            Collection<NewTopic> topics = Collections.singletonList(new NewTopic(topicName, 1, (short) 1));
+
+            Awaitility
+                .await()
+                .untilAsserted(() -> {
+                    assertThatThrownBy(() -> adminClient.createTopics(topics).all().get(30, TimeUnit.SECONDS))
+                        .hasCauseInstanceOf(SaslAuthenticationException.class);
+                });
+        }
+    }
+
+    private static String getJaasConfig() {
+        String jaasConfig =
+            "org.apache.kafka.common.security.plain.PlainLoginModule required " +
+            "username=\"admin\" " +
+            "password=\"admin\" " +
+            "user_admin=\"admin\" " +
+            "user_test=\"secret\";";
+        return jaasConfig;
+    }
+
+    private void testKafkaFunctionality(String bootstrapServers) throws Exception {
+        testKafkaFunctionality(bootstrapServers, false, 1, 1);
+    }
+
+    private void testSecureKafkaFunctionality(String bootstrapServers) throws Exception {
+        testKafkaFunctionality(bootstrapServers, true, 1, 1);
+    }
+
+    private void testKafkaFunctionality(String bootstrapServers, boolean authenticated, int partitions, int rf)
+        throws Exception {
+        ImmutableMap<String, String> adminClientDefaultProperties = ImmutableMap.of(
+            AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG,
+            bootstrapServers
+        );
+        Properties adminClientProperties = new Properties();
+        adminClientProperties.putAll(adminClientDefaultProperties);
+
+        ImmutableMap<String, String> consumerDefaultProperties = ImmutableMap.of(
+            ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
+            bootstrapServers,
+            ConsumerConfig.GROUP_ID_CONFIG,
+            "tc-" + UUID.randomUUID(),
+            ConsumerConfig.AUTO_OFFSET_RESET_CONFIG,
+            "earliest"
+        );
+        Properties consumerProperties = new Properties();
+        consumerProperties.putAll(consumerDefaultProperties);
+
+        ImmutableMap<String, String> producerDefaultProperties = ImmutableMap.of(
+            ProducerConfig.BOOTSTRAP_SERVERS_CONFIG,
+            bootstrapServers,
+            ProducerConfig.CLIENT_ID_CONFIG,
+            UUID.randomUUID().toString()
+        );
+        Properties producerProperties = new Properties();
+        producerProperties.putAll(producerDefaultProperties);
+
+        if (authenticated) {
+            adminClientProperties.putAll(this.properties);
+            consumerProperties.putAll(this.properties);
+            producerProperties.putAll(this.properties);
+        }
+        try (
+            AdminClient adminClient = AdminClient.create(adminClientProperties);
+            KafkaProducer<String, String> producer = new KafkaProducer<>(
+                producerProperties,
                 new StringSerializer(),
                 new StringSerializer()
             );
             KafkaConsumer<String, String> consumer = new KafkaConsumer<>(
-                ImmutableMap.of(
-                    ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
-                    bootstrapServers,
-                    ConsumerConfig.GROUP_ID_CONFIG,
-                    "tc-" + UUID.randomUUID(),
-                    ConsumerConfig.AUTO_OFFSET_RESET_CONFIG,
-                    "earliest"
-                ),
+                consumerProperties,
                 new StringDeserializer(),
                 new StringDeserializer()
             );

